### ISSUE TITLE
Adding a viewer to R-Instat

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/Instat#2319]
When we demonstrated R-Instat there was a request for an additional sheet - as a viewer.  I now understand this could be trivial one way, namely to invoke the R view command.  Danny demonstrated this and it looks fun and also another "free gift" from R.

It can be invoked, in R-Instat through the calculate dialogue - just to see what it does!  Then I suggest as follows:

a) It becomes a right-click option from the tabs at the bottom of the grid.  This does not go to a dialogue, but just opens the view on that sheet.
b) It also could be an extra option on the **Organise > Data Frame > View Data** dialogue.  That might give access to some options, e.g. to just view the selected columns, and perhaps not all the rows, though it does seem fast!

I hope this could be quite simple and also fun for someone?  I am suggesting Steven, but am also copying Francois as a route to sharing the discussion with the others.

Roger

**Comment by FrancoisJRenaud** [comment]
Hi @rdstern, 
In order to make sure @stevenndungu  and I get you, what do you exactly mean by "viewer". Is that a specific R command that produces a "view" on a given sheet or part of a sheet ? 
Found this on the net: https://stat.ethz.ch/R-manual/R-devel/library/utils/html/View.html
There is indeed a command View in the package utils.
Are you thinking of that one ?
![new picture](https://cloud.githubusercontent.com/assets/22272110/20638972/ca261280-b3c7-11e6-8d0b-3edf958a4542.jpg)
Then you would like that view to appear in a separate window, like the graphs did so far ? Or should we include the view in the output window like the graphs will be doing soon ?
Cheers, 
François

**Comment by rdstern** [comment]
I was thinking of the R view command - at least that is what Danny showed me is easy to do.  

And I was thinking of it as a separate viewer, like the graphs are/were.  This could allow us to look at 2 sheets together, one in our usual grid and one in the viewer.

Later we may also want to be able to view data in the output window, but that will be different.  That might make use of stargazer, or the sjmisc/sjplot I guess.

**Comment by FrancoisJRenaud** [comment]
Ok, @dannyparsons  is it the same "View" command as the one I found above ? In utils ?
Otherwise I ve been testing the stargazer in the new output window and it works well. So this might be near future... I ve just sent an email summarizing the progress.

**Comment by dannyparsons** [comment]
Yes it's simply the `View` command. So it should be very simple to add an option on the tabs right click to simply call the `View` command on the selected data frame.
Then another simple job is to update the View dialog so that there is an option to view the data frame in a separate viewer, as well as in the output window (which is currently the only option).

How we display things in the output window is a separate discussion, where stargazer is one possibility, but the changes here can be done independently of that decision.

**Comment by rdstern** [comment]
I agree with Danny here.  I am still keen on this simple way of being able to view a sheet in parallel to the use of our grid - where we only see one sheet at a time.  So, for example if we wanted to consider merging data from 2 sheets, then we could look at them together first.  Or, if we have stacked, then we could look at the sheet with the unstacked data and the stacked data together.  So I see this as another Window - and we might be able also to invoke it through the View menu.

Separately from this feature, I am also keen to be able to improve on the current display of data in the output window, which currently uses the **Organise > Data Frame > View** dialogue.  But that's a discussion that fits better in the issue where you are discussing the options we now have for that window.  And that is exciting too.

**Comment by stevenndungu** [comment]
Thanks Roger, Danny and Francois for the discussions above. You have made it clear, am happy to start  working on it immediately. Its a command I use so often in R... and now seeing it  in R-Instat would be really nice.

**Comment by rdstern** [comment]
I assume the work on the viewer has started?  I now find that the old view dialogue doesn't work any more.  It now has a default of 0 rows to display, which I can't change.  I am looking forward to the right-click on the bottom tab being added.

**Comment by stevenndungu** [comment]
Yes am done working on it and its now working well. Currently awaiting review and consequently merging.

**Comment by rdstern** [comment]
Congratulations on the viewer, which is working well from the right-click menu at the bottom.  Great.

a) Could you look at the title for the frame and make it more friendly.
b) However the window "goes off" too quickly!  I suggest it should stay visible until you choose to close it.  This is a general issue with the windows , e.g. also for the View > Column Metadata etc.  In fact it is better than the metadata, because it is still accessible from the icon at the bottom - like a graph.  Could this issue please be discussed with Danny?
c) One use of this facility is to view 2 different data frames at the same time. I can't yet do this.  Whenever I click to move to a different data frame (or indeed anywhere outside the viewer, it disappears.  That makes it much less useful than it should be.
d) For example I have a climatic data file and a station geography file.  I would like to view them together.  How can I do this? 

e) Perhaps unrelated but now the old view dialogue **Organise > Data Frame > View** is now still not working, as I mentioned above.  That used to work and was a nice simple dialogue.  Now the up-down control is stuck on 0 and I also can not type into that control - perhaps that isn't intended.
f) Earlier we thought that your new viewer could alternatively be called from this dialogue.  I don't mind either way.  But I would at least like our old **Organise > Data Frame > View** to work as before.  
g) Could the viewer also be available from the View menu?  It is a similar sort of window.  There would need to be a "current" sheet for this to be possible.  If not, then I suggest it would become a bit more important to also have it available from the **Organise > Data Frame > View** dialogue.

**Comment by dannyparsons** [comment]
b) The window should behave like a graph window, where it starts minimised. Currently we don't have any control over that. Like with the graphics we would need to do "our own" viewer to get it how we like. The other windows like Metadata are inside the software so we could do something about those.
c, d) This is because it's a completely separate window. It's like having another program on top of R-Instat, you can't see both at the same time. The way you could do this is to resize R-Instat to the left side of the screen and the viewer on the right side. Then you can see both together, but it's not ideal.

The other comments I believe @stevenndungu is working to fix them as expect, but if not he can comment.

**Comment by rdstern** [comment]
Yes, and as it is separate, it does remain accessible until closed - or (I presume) overwritten.  Oh no.  That's neat, you can open more than one - and see them together.  Fun!

**Comment by stevenndungu** [comment]
@rdstern 
Thanks for the kind comments above. Here are changes I have made in regard to your comments above;
a) The title of the data frame to viewed is now derived from the name of the data frame to be viewed which is friendly.
e) - the up-down control is now working properly as before.
-Viewer option is as well added in the Organise > Data Frame > View dialogue. This is in such away that we have two radio buttons 1) For the output to be displayed in the Output Window as was before and 2) For the output to be displayed on a separate window (the Viewer window)...this was in consultation with Danny. 
Steven

**Comment by stevenndungu** [comment]
Yes, it is yet to be merged.

**Comment by rdstern** [comment]
OK, now I see the changes.  Two points so far:

1. The alternative radio buttons are not obvious alternatives to me.  If this stays as radio buttons, then I suggest we change the layout (and possibly the default).  If we have the "Display in Separate Window" first, then the 2 alternatives will be obvious.
 2.  Have you tested the "Display in Output Window" fully on your machine?  I wonder if you have spent most of your time on making sure the new bit you added is working well - which I think it is.  And less time on testing the other part - which used to work!  Here is what I found.  I was using the usual survey dataset with 36 rows.

1.  the default in the up-down is still 0 rows, which is rarely needed!
2.  the up-down still does not seem to work.  But I can type into the control now.  So I typed 6 and chose all the columns.  I still got nothing displayed.  It had effectively put itself back to 0.
3.  When I went back to the dialogue I could no longer type into the field.  It remains stubbornly at 0.
